include/envoy/network/io_handle.h

```diff
@@ -223,6 +228,18 @@ class IoHandle {
    * @return peer's address as @ref Address::InstanceConstSharedPtr
    */
   virtual Address::InstanceConstSharedPtr peerAddress() PURE;
+
+  /**
+   * Creates a file event that will signal when the io handle is readable, writable or closed.
+   * @param dispatcher dispatcher to be used to allocate the file event.
+   * @param cb supplies the callback to fire when the handle is ready.
+   * @param trigger specifies whether to edge or level trigger.
+   * @param events supplies a logical OR of @ref Event::FileReadyType events that the file event
+   *               should initially listen on.
+   * @return @ref Event::FileEventPtr
+   */
+  virtual Event::FileEventPtr createFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
```

Comment:
The long term plan continues to be to completely remove "fd" from all top level dispatcher interfaces, right? My main concern is there is nothing right now that prevents not calling the original FileEvent factory function with the IoHandle fd, right? Can we add relevant comments around usage, TODOs for eventual removal, warnings, etc.? My quick take would be that if we have a similar function on the filesystem interface we might be able to completely remove the public interface that allows file event creation from fd sooner rather than later (I'm not sure what else actually uses that factory currently)? wdyt?

Reply:
That's exactly right. The goal is to use IoHandles instead of fds whenever possible. 

Apart from the sockets, others that use the dispatcher's original `FileEvent` factory function are:
- `WatcherImpl`, which creates inotify events using raw fds 
- `DnsImpl`, which relies on raw fds from Ares
- `HotRestartingParent`, which also uses raw fds

In theory, we can change all of these, including `IoSocketHandleImpl` to allocate `FileEventPtr`s directly and then completely remove the dispatcher's public interface. The only thing we'll miss out on is the `isThreadSafe()` check, but maybe we can make the function public? I can take care of this but should we do it as part of this PR or a separate one?

Finally, out of those 3, maybe we should update at least `DnsImpl` to use `IoHandles` (not sure if that's easy).

Reply:
> In theory, we can change all of these, including IoSocketHandleImpl to allocate FileEventPtrs directly and then completely remove the dispatcher's public interface. The only thing we'll miss out on is the isThreadSafe() check, but maybe we can make the function public? I can take care of this but should we do it as part of this PR or a separate one?

I would prefer we just do it now if you are up for it, but if you feel strongly we can do it as a fast follow up?

> Finally, out of those 3, maybe we should update at least DnsImpl to use IoHandles (not sure if that's easy).

Yeah that would be nice. Maybe add a TODO there for follow up later?

Reply:
Was about to start work on it but it turns out `FileEventImpl` depends on `DispatcherImpl` and all is bundled in `//source/common/event:dispatcher_lib`. Trying to add the latter as dependency to the `default_socket_interface_lib` quickly runs into circular dependencies. I guess this is because we don't want to expose the scheduler's implementation, i.e., libevent's `event_base`, in the dispatcher's interface. 

Unless there's a smarter way, we'll have have to separate `FileEventImpl` from `DispatcherImpl` and that might not be trivial (they'd have to exchange libevent specific objects over generic apis). The alternative would be to still use the dispatcher as a `FileEvent` factory and work towards removing the fd from `IoHandle` (more of that [here](https://github.com/florincoras/envoy/commit/9b42b933f1e143afcac1835587b47af72bf4f249) and #12736). But that exercise is also not trivial because "higher level transports" like ssl and quic do want to see the raw fds. So it might be some time until we cleanup everything. 

What option would you prefer?

Reply:
Eliminating use of the IoHandle::fd() method seems like progress.

Direct creation of FileEvents instead of going through the dispatcher would reduce out ability to mock out FileEvent operations or inject test versions that can easily be marked read/write blocked, unless we add some factory interface similar to the one provided by the dispatchers today.

Reply:
Me too. After this PR, 12736 and [this](https://github.com/florincoras/envoy/commit/2f7a002b55c84ed5bc110ff8830df2ca3a4ad42a) are merged, we should be left with tls/quic/maybe some udp code that touch the fd(). Maybe be can make the fd() function _fd() and insist that people do not use it in reviews or friend classes?

Reply:
Yes potentially we can just have a format/lint checker while we do further cleanups.

Reply:
and yes maybe also rename `fdDoNotUse()` or something until we fix it all.

Reply:
Works for me! Can we do it after all the PRs are merged to limit impact, i.e., change name and then delete them?

Reply:
Yup that's fine.

Reply:
Thanks! Added to my todo list :-)